internal-dns-client/src/multiclient.rs

```diff
@@ -0,0 +1,651 @@
+                        .await?;
+                }
+            };
+        }
+        Ok(())
+    }
+
+    // Utility function to insert:
+    // - A set of uniquely-named AAAA records, each corresponding to an address
+    // - An SRV record, pointing to each of the AAAA records.
+    async fn insert_dns_records_internal(
+        &self,
+        aaaa: Vec<(crate::names::AAAA, SocketAddrV6)>,
+        srv_key: crate::names::SRV,
+    ) -> Result<(), DnsError> {
+        let mut records = Vec::with_capacity(aaaa.len() + 1);
+
+        // Add one DnsKv per AAAA, each with a single record.
+        records.extend(aaaa.iter().map(|(name, addr)| DnsKv {
+            key: DnsRecordKey { name: name.to_string() },
+            records: vec![DnsRecord::Aaaa(*addr.ip())],
+        }));
+
+        // Add the DnsKv for the SRV, with a record for each AAAA.
+        records.push(DnsKv {
+            key: DnsRecordKey { name: srv_key.to_string() },
+            records: aaaa
+                .iter()
+                .map(|(name, addr)| {
+                    DnsRecord::Srv(Srv {
+                        prio: 0,
+                        weight: 0,
+                        port: addr.port(),
+                        target: name.to_string(),
+                    })
+                })
+                .collect::<Vec<_>>(),
+        });
+        self.dns_records_set(&records).await
+    }
+
+    /// Sets a records on all DNS servers.
+    ///
+    /// Returns an error if setting the record fails on any server.
+    pub async fn dns_records_set<'a>(
+        &'a self,
+        body: &'a Vec<crate::types::DnsKv>,
+    ) -> Result<(), DnsError> {
+        stream::iter(&self.clients)
+            .map(Ok::<_, DnsError>)
+            .try_for_each_concurrent(None, |client| async move {
+                client.dns_records_set(body).await?;
+                Ok(())
+            })
+            .await?;
+
+        Ok(())
+    }
+
+    /// Deletes records in all DNS servers.
+    ///
+    /// Returns an error if deleting the record fails on any server.
+    pub async fn dns_records_delete<'a>(
+        &'a self,
+        body: &'a Vec<crate::types::DnsRecordKey>,
+    ) -> Result<(), DnsError> {
+        stream::iter(&self.clients)
+            .map(Ok::<_, DnsError>)
+            .try_for_each_concurrent(None, |client| async move {
+                client.dns_records_delete(body).await?;
+                Ok(())
+            })
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum ResolveError {
+    #[error(transparent)]
+    Resolve(#[from] trust_dns_resolver::error::ResolveError),
+
+    #[error("Record not found for SRV key: {0}")]
+    NotFound(crate::names::SRV),
+}
+
+/// A wrapper around a DNS resolver, providing a way to conveniently
+/// look up IP addresses of services based on their SRV keys.
+pub struct Resolver {
+    inner: Box<TokioAsyncResolver>,
```

Comment:
What's the motivation for boxing this?

Reply:
I'm using it in subsequent patches in an enum (you can provide an IP from a config, or use a resolver), and Clippy complained that it was too large (500+ bytes). Since it's internal to the struct, it shouldn't matter for users of this API.